pygpcca/_gpcca.py

```diff
@@ -1001,14 +1013,19 @@ def optimize(
             chi_list.append(chi)
             rot_matrix_list.append(rot_matrix)
 
-        opt_idx = np.argmax(crispness_list)
+        if np.any(np.array(crispness_list) > 0.0):
+            opt_idx = np.argmax(crispness_list)
+        else:
+            raise ValueError("Clustering wasn't successful. Try different cluster numbers.")
         self._m_opt = min(m_list) + opt_idx
         self._chi = chi_list[opt_idx]
         self._rot_matrix = rot_matrix_list[opt_idx]
-        self._crispness = crispness_list[opt_idx]
+        self._crispness = np.array(crispness_list)
+        self._crispness_opt = crispness_list[opt_idx]
         self._X = self._p_X[:, : self._m_opt]
         self._R = self._p_R[: self._m_opt, : self._m_opt]
-        self._eigenvalues = self._p_eigenvalues[: self._m_opt]
+        self._top_eigenvalues = self._p_eigenvalues[: self._m_opt]
```

Comment:
Top eigenvalues are the ones up to the optimal cluster number?

Reply:
Exactly!